packages/flutter/lib/src/services/spell_check.dart

```diff
@@ -0,0 +1,86 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'package:flutter/foundation.dart';
+
+/// A data structure representing a range of misspelled text and the suggested
+/// replacements for this range. For example, one [SuggestionSpan] of the
+/// [List<SuggestionSpan>] suggestions of the [SpellCheckResults] corresponding
+/// to "Hello, wrold!" may be:
+/// ```dart
+/// SuggestionSpan(7, 11, List<String>.from["word, world, old"])
+/// ```
+@immutable
+class SuggestionSpan {
+  /// Creates a span representing a misspelled range of text and the replacements
+  /// suggested by a spell checker.
+  ///
+  /// The [startIndex], [endIndex], and replacement [suggestions] must all not
+  /// be null.
+  const SuggestionSpan(this.startIndex, this.endIndex, this.suggestions)
+      : assert(startIndex != null),
+        assert(endIndex != null),
+        assert(suggestions != null);
+
+  /// The start index of the misspelled range of text.
+  final int startIndex;
+
+  /// The end index of the misspelled range of text, inclusive.
+  final int endIndex;
+
+  /// The alternate suggestions for misspelled range of text.
+  final List<String> suggestions;
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other))
+        return true;
+    return other is SuggestionSpan &&
+        other.startIndex == startIndex &&
+        other.endIndex == endIndex &&
+        listEquals<String>(other.suggestions, suggestions);
+  }
+
+  @override
+  int get hashCode => Object.hash(startIndex, endIndex, hashList(suggestions));
+}
+
+/// A data structure grouping together the [SuggestionSpan]s and related text of
+/// results returned by a spell checker.
+@immutable
+class SpellCheckResults {
+  /// Creates results based off those received by spell checking some text input.
+  const SpellCheckResults(this.spellCheckedText, this.suggestionSpans);
+
+  /// The text that the [suggestionSpans] correspond to.
+  final String spellCheckedText;
+
+  /// The spell check results of the [spellCheckedText].
+  /// See also:
+  ///
+  ///  * [SuggestionSpan], the ranges of misspelled text and corresponding
+  ///    replacement suggestions.
+  final List<SuggestionSpan> suggestionSpans;
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other))
+        return true;
+    return other is SpellCheckResults &&
+        other.spellCheckedText == spellCheckedText &&
+        listEquals<SuggestionSpan>(other.suggestionSpans, suggestionSpans);
+  }
+
+  @override
+  int get hashCode => Object.hash(spellCheckedText, hashList(suggestionSpans));
+}
+
+/// Determines how spell check results are received for text input.
+abstract class SpellCheckService {
+  /// Facilitates a spell check request.
+  Future<SpellCheckResults?> fetchSpellCheckSuggestions(
```

Comment:
the `spellCheckedText` in the returned `SpellCheckResults` will be the same as the input `text`? If so should we just return a list of `SuggestionSpan`s?

Reply:
I believe the answer to this question is guaranteed to be "yes" based on the way the engine PR was implemented in the end.  So SuggestionSpans make sense to me.

Reply:
This would mean I would just save the text (saying something like `final text = _value.text` and pair it with the results within `EditableText` [here](https://github.com/flutter/flutter/pull/96629/files#diff-f5f93c879cef9a102adbf148583de9f3d9a05f9678fdcad7bde106f62200bb34R2583) if the results are not null, but that would defeat the purpose of having the engine send back the text that the results are related to. I remember bringing this up before. Do either of you remember the argument against doing this?

Reply:
I think at that time we had no guarantee that the results being sent back were for the most recent `text`, but now we ensured that by handling the requests one at a time in the engine, right?  If so then it seems fine to just save the text like you described, and indeed there is no point in sending the text back from the engine anymore?

Reply:
Ah gotcha. Ok, I'll make this change and after this PR looks good, I'll make a PR for the engine change.